### PR TITLE
Of review

### PR DESCRIPTION
**Pull request by oliverfleetwood**
I will add comments and suggestions for edits here as a pull request. No need to merge it, it is mostly to be used as material for discussion

### CODE REVIEW

File: modules/feature_extraction/pca_feature_extractor.py [hunk 1]
```diff
@@ -17,10 +17,12 @@
 
 class PCAFeatureExtractor(FeatureExtractor):
 
-    def __init__(self, samples, cluster_indices, n_splits=10, scaling=True, filter_by_distance_cutoff=False, contact_cutoff=0.5, n_components=None, name="PCA"):
+    def __init__(self, samples, cluster_indices, n_splits=10, scaling=True, filter_by_distance_cutoff=False, contact_cutoff=0.5,
+                 variance_cutoff=0.75, n_components=None, name="PCA"):
 
         FeatureExtractor.__init__(self, samples, cluster_indices, n_splits=n_splits, n_iterations=1, scaling=scaling, filter_by_distance_cutoff=filter_by_distance_cutoff, contact_cutoff=contact_cutoff, name=name)
         self.n_components = n_components
+        self.variance_cutoff = 0.75
```
Comment (oliverfleetwood) [inline modules/feature_extraction/pca_feature_extractor.py]: making the variance configurable

File: modules/feature_extraction/random_forest_feature_extractor.py [hunk 2]
```diff
@@ -25,7 +25,9 @@ def __init__(self, samples, cluster_indices, n_splits=10, n_iterations=10, scali
 
     def train(self, train_set, train_labels):
         # Construct and train classifier
-        classifier = RandomForestClassifier(n_estimators=self.n_estimators, n_jobs=self.njobs)
+        classifier = RandomForestClassifier(n_estimators=self.n_estimators, 
+                                            n_jobs=self.njobs,
```
Comment (oliverfleetwood) [inline modules/feature_extraction/random_forest_feature_extractor.py]: Making it possible to use a fixed seed for reproducability, just as for MLP, and settings njobs to use all processors

File: modules/visualization.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def vis_projected_data(proj_data, cluster_indices, fig, title):
         for i in range(n_dims):
             for j in range(i+1,n_dims):
                 axes.append(fig.add_subplot(np.ceil(n_combi/3),3,counter))
-                axes[counter-1].scatter(proj_data[:,i],proj_data[:,j],s=15,c=cluster_indices,edgecolor='')
+                axes[counter-1].scatter(proj_data[:,i],proj_data[:,j],s=15,c=cluster_indices,edgecolor='', alpha=0.3)
```
Comment (oliverfleetwood) [inline modules/visualization.py]: Should make it easier to see when clusters are on top of eachother

File: modules/feature_extraction/kl_divergence_feature_extractor.py [hunk 4]
```diff
@@ -19,20 +19,22 @@
...
 
     def train(self, data, labels):
-        n_clusters = labels.shape[1]
 
-        self.feature_importances = np.zeros((n_clusters, data.shape[1]))
-        for i_cluster in range(n_clusters):
-            data_cluster = data[labels[:, i_cluster] == 1, :]
-            data_rest = data[labels[:, i_cluster] == 0, :]
-            self.feature_importances[i_cluster, :] = self._KL_divergence(data_cluster[:,:,np.newaxis], data_rest[:,:,np.newaxis])
+        if self.cluster_split_method == "one_vs_rest":
```
Comment (oliverfleetwood) [inline modules/feature_extraction/kl_divergence_feature_extractor.py]: suggestion to make this method configurable for easier comparison

File: modules/feature_extraction/kl_divergence_feature_extractor.py [hunk 5]
```diff
@@ -42,8 +44,9 @@ def _KL_divergence(self, x, y):
         if self.bin_size is None:
             std = np.zeros(n_features)
             for i_feature in range(n_features):
-                std[i_feature] = np.std(x[:, i_feature])
-            self.bin_size = np.mean(std)
+                std[i_feature] = np.std(x[:, i_feature]) 
+            self.bin_size = np.mean(std) #Can be done with x.std(axis=X).mean() instead of loop here.
+            # As long as we use MaxMinScaler it maxes sense to have fixed bindwith, or?
```
Comment (oliverfleetwood) [inline modules/feature_extraction/kl_divergence_feature_extractor.py]: As long as we use MaxMinScaler it maxes sense to have fixed bindwith, or?

File: modules/feature_extraction/kl_divergence_feature_extractor.py [hunk 6]
```diff
@@ -55,16 +58,24 @@ def _KL_divergence(self, x, y):
                 bin_n = int((bin_max - bin_min) / self.bin_size)
                 x_prob = np.histogram(x[:, i_feature], bins=bin_n, range=(bin_min, bin_max), density=True)[0] + 1e-9
                 y_prob = np.histogram(y[:, i_feature], bins=bin_n, range=(bin_min, bin_max), density=True)[0] + 1e-9
-                DKL[i_feature] = 0.5 * (entropy(x_prob, y_prob) + entropy(y_prob, x_prob))
+                DKL[i_feature] = 0.5 * (entropy(x_prob, y_prob) + entropy(y_prob, x_prob)) #Is mean the way to go? What about using max or min?
```
Comment (oliverfleetwood) [inline modules/feature_extraction/kl_divergence_feature_extractor.py]: use max instead of avg?

**Comment by oliverfleetwood** [comment]
@anniewesterlund can we look at dataprojection together? I could use a walkthrough.

**Comment by anniewesterlund** [comment]
Sure! Let’s do that when both are there.
tis 11 dec. 2018 kl. 18:12 skrev Oliver Fleetwood <notifications@github.com

**Pull request merged**